src/modules/networks/networks-store.ts

```diff
@@ -1,101 +1,170 @@
-      this.setState({ networks });
-      this.needsRetry = networksValue.status === 'rejected';
-      return networks;
+  private async updateNetworks() {
+    const savedChainConfigs = await this.getEthereumChainConfigs?.();
+    const networks = new Networks({
+      networks: mergeNetworkConfigs(
+        this.networkConfigs,
+        this.customNetworkConfigs
+      ),
+      ethereumChainConfigs: savedChainConfigs || [],
     });
+    this.setState({ networks });
+    return networks;
+  }
+
+  private async fetchNetworks({
+    chains,
+    update,
+  }: {
+    chains?: string[];
+    update?: boolean;
+  }) {
+    const existingIdsSet = this.getKnownIdsSet();
+    const shouldUpdateNetworksInfo =
+      update || chains?.some((id) => !existingIdsSet.has(id));
+    const existingNetworks = this.getState().networks;
+    if (!shouldUpdateNetworksInfo && existingNetworks) {
+      return existingNetworks;
+    }
+
+    const savedChainConfigs = await this.getEthereumChainConfigs?.();
+    const savedIds = savedChainConfigs?.map((config) => config.id);
+    const chainsToFetch = Array.from(
+      new Set([...(savedIds || []), ...(chains || [])])
+    );
+
+    const [extraNetworkConfigs, commonNetworkConfigs] =
+      await Promise.allSettled([getNetworks(chainsToFetch), getNetworks()]);
+    const updatedNetworkConfigs = mergeNetworkConfigs(
+      commonNetworkConfigs.status === 'fulfilled'
+        ? commonNetworkConfigs.value
+        : [],
+      extraNetworkConfigs.status === 'fulfilled'
+        ? extraNetworkConfigs.value
+        : []
+    );
+
+    this.networkConfigs = mergeNetworkConfigs(
+      this.networkConfigs,
+      updatedNetworkConfigs
+    );
+    const fulfilledNetworkIdSet = new Set(
+      this.networkConfigs.map(({ id }) => id)
+    );
+    this.customNetworkConfigs =
+      savedChainConfigs
+        ?.filter((config) => !fulfilledNetworkIdSet.has(config.id))
+        .map((config) => toNetworkConfig(config.value)) || [];
+
+    return this.updateNetworks();
+  }
+
+  private async fetchNetworkById(chainId: ChainId) {
+    const shouldUpdateNetworksInfo = [
+      ...this.networkConfigs,
+      ...this.customNetworkConfigs,
+    ].every((network) => Networks.getChainId(network) !== chainId);
+    const existingNetworks = this.getState().networks;
+    if (!shouldUpdateNetworksInfo && existingNetworks) {
+      return existingNetworks;
+    }
+    const network = await getNetworkByChainId(chainId);
+    if (network) {
+      this.networkConfigs = mergeNetworkConfigs(this.networkConfigs, [network]);
+    }
+    return this.updateNetworks();
   }
 
-  async load(): Promise<Networks> {
-    if (!this.loaderPromise) {
-      this.loaderPromise = this.fetchAndUpdate();
+  async pushConfigs(...extraNetworkConfigs: NetworkConfig[]) {
+    this.networkConfigs = mergeNetworkConfigs(
+      this.networkConfigs,
+      extraNetworkConfigs
+    );
+    return this.updateNetworks();
+  }
+
+  async load(chains?: string[]) {
```

Comment:
I don't like that we create lots of promise objects even if all chains are known already

Reply:
Let me remove this for real and check it properly again then :)